Experimental Feature: Provide Unix stat information in filesystem output

JamesWTruher opened the pull request:
# PR Summary

This PR provides new behavior, as experimental feature `PSUnixFileStat`, to include data from the Unix stat API in the output of the file system provider to facilitate a more Unix-like file listing. It adds a new note property in the filesystem provider named `UnixStat` which includes a common expression of the `stat(2)` API from the underlying Unix type system. The native calls in the `powershell-native` assembly have created 

The output from `Get-ChildItem` should look something like this:

```powershell
PS> dir | select -first 4 -skip 5


    Directory: /Users/jimtru/src/github/forks/JamesWTruher/PowerShell-1

UnixMode   User             Group                 LastWriteTime           Size Name
--------   ----             -----                 -------------           ---- ----
drwxr-xr-x jimtru           staff              10/23/2019 13:16            416 test
drwxr-xr-x jimtru           staff               11/8/2019 10:37            896 tools
-rw-r--r-- jimtru           staff               11/8/2019 10:37         112858 build.psm1
-rw-r--r-- jimtru           staff               11/8/2019 10:37         201297 CHANGELOG.md
```

The username and groupname are retrieved at the time of formatting to reduce the number of PInvokes required.

## PR Context

I've been somewhat frustrated with viewing the output of `Get-ChildItem` since it's different from `/bin/ls -l` output. This is an attempt to improve the experience with regard to the file system provider when running on a non-Windows platforms.

The old format is still available by selecting the `children` view with format-table:

```powershell
PS> dir | select -skip 5 -first 4 | ft -view children

    Directory: /Users/jimtru/src/github/forks/JamesWTruher/PowerShell-1

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          10/23/2019  1:16 PM                test
d----           11/8/2019 10:37 AM                tools
-----           11/8/2019 10:37 AM         112858 build.psm1
-----           11/8/2019 10:37 AM         201297 CHANGELOG.md
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

JamesWTruher committed:
Checkpoint for UnixStat code

JamesWTruher committed:
revert classes to internal

JamesWTruher committed:
Add tests for UnixStat

JamesWTruher committed:
UnixStat is only available for Unix platforms

JamesWTruher committed:
update count of available types (now that I added an additional format for UnixStat

JamesWTruher committed:
Update to use preview.4 of the native library

It includes the new common Unix stat methods.

PoshChan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -561,20 +561,255 @@ internal static int NonWindowsGetProcessParentPid(int pid)
         // the string perror would give you for the last set value of errno.
```
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
UnixFileSystem additions.Other properties of UnixStat object.Error occurred in Context block
```powershell
The term '/bin/ls' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\UnixStat.Tests.ps1:98 char:18
```

PoshChan replied:
@JamesWTruher, your last commit had 3 failures in `PowerShell-CI-macos`
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the user name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 76
76:             $user = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[2]

```
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the group name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 82
82:             $group = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[3]

```
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 3 failures in `PowerShell-CI-linux`
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the user name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 76
76:             $user = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[2]

```
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the group name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 82
82:             $group = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[3]

```
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 3 failures in `PowerShell-CI-macos`
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the user name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 76
76:             $user = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[2]

```
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the group name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 82
82:             $group = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[3]

```
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 3 failures in `PowerShell-CI-linux`
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the user name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 76
76:             $user = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[2]

```
UnixFileSystem additions.Validation of additional properties on file system objects.Should retrieve the group name for the file
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1: line 82
82:             $group = (/bin/ls -ld /tmp/foo).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[3]

```
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

JamesWTruher committed:
Skip BeforeAll on Windows

JamesWTruher committed:
Fix dumb typo and test against proper file

JamesWTruher committed:
Round 1 of CodeFactor fixes

PoshChan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,285 @@ internal static int NonWindowsGetProcessParentPid(int pid)
             return IsMacOS ? Unix.NativeMethods.GetPPid(pid) : Unix.GetProcFSParentPid(pid);
```
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

JamesWTruher committed:
Code Factor fixes round 2

PoshChan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,288 @@ internal static int NonWindowsGetProcessParentPid(int pid)
             return IsMacOS ? Unix.NativeMethods.GetPPid(pid) : Unix.GetProcFSParentPid(pid);
```
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

JamesWTruher committed:
Code Factor fixes round 3

adityapatwardhan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,289 @@ internal static int NonWindowsGetProcessParentPid(int pid)
+                // helper for getting unix mode
+                private Dictionary<StatMask, string> modeMap = new Dictionary<StatMask, string>()
+                {
+                        { StatMask.OwnerRead, "r" },
```
Maybe a slight optimization, if you declare "r", "w" and "x" as constants, then they can be reused.

JamesWTruher replied:
done!

adityapatwardhan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,289 @@ internal static int NonWindowsGetProcessParentPid(int pid)
+                /// <returns>The mode converted into a Unix style string similar to the output of ls.</returns>
+                public string GetModeString()
+                {
+                    StatMask[] permissions = new StatMask[]
```
Can this be defined statically? would reduce allocations i think.

adityapatwardhan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,289 @@ internal static int NonWindowsGetProcessParentPid(int pid)
+                    };
+
+                    // start the mode string with the ItemType
+                    System.Text.StringBuilder sb = new System.Text.StringBuilder();
```
The length of the output string is fixed to 10 i guess, so we should initialize the string builder with 10 as the capacity.

JamesWTruher replied:
i rewrote this method using a char array.

PoshChan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,289 @@ internal static int NonWindowsGetProcessParentPid(int pid)
             return IsMacOS ? Unix.NativeMethods.GetPPid(pid) : Unix.GetProcFSParentPid(pid);
```
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-macos`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-linux`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 4, but got collection with size 5 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 4
```

PoshChan replied:
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 5, but got collection with size 4 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 5
```

adityapatwardhan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -644,21 +992,39 @@ internal static class NativeMethods
+                    internal int tm_isdst;
                 }
 
+                // We need a way to convert a DateTime to a unix date.j
```
```suggestion
                // We need a way to convert a DateTime to a unix date.
```

JamesWTruher replied:
fixed!

adityapatwardhan commented on src/System.Management.Automation/engine/TypeTable_Types_Ps1Xml.cs:
```diff
@@ -9198,6 +9198,53 @@ private void Process_Types_Ps1Xml(string filePath, ConcurrentBag<string> errors)
+
+            if (ExperimentalFeature.IsEnabled("PSUnixFileStat"))
+            {
+            typeName = @"System.IO.FileSystemInfo";
```
formatting looks off.

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1:
```diff
@@ -0,0 +1,127 @@
+        $skipTest = $true
+        if ( Get-ExperimentalFeature -Name $experimentalFeatureName ) {
+            $skipTest = $false
+        }
```
This can be:

$skipTest = -not $EnabledExperimentalFeatures.Contains('PSUnixFileStat')

JamesWTruher replied:
fixed!

JamesWTruher committed:
We now have 5 views for System.IO.FileInfo with the UnixStat experimental feature

JamesWTruher committed:
Rewrite of GetModeString to not use stringbuilder, but start with an array of chars

A number of smaller fixes suggested by review.

PoshChan commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,283 @@ internal static int NonWindowsGetProcessParentPid(int pid)
             return IsMacOS ? Unix.NativeMethods.GetPPid(pid) : Unix.GetProcFSParentPid(pid);
```
@JamesWTruher, your last commit had 1 failures in `PowerShell-CI-windows`
Get-FormatData.Can get format data requiring '-PowerShellVersion 5.1'
```powershell
Expected a collection with size 5, but got collection with size 4 @(System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition, System.Management.Automation.FormatViewDefinition).
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Get-FormatData.Tests.ps1: line 18
18:         $format.FormatViewDefinition | Should -HaveCount 5
```

JamesWTruher committed:
The count of format views for System.IO.FileInfo is only 5 on non-Windows

JamesWTruher committed:
Code Factor changes 3

rjmholt commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -589,19 +847,105 @@ public static bool IsHardLink(FileSystemInfo fs)
+                    cs.GroupId = css.GroupId;
+                    cs.HardlinkCount = css.HardlinkCount;
+                    cs.Size = css.Size;
+                    cs.AccessTime = new DateTime(1970, 1, 1).AddSeconds(css.AccessTime).ToLocalTime();
```
```suggestion
                    cs.AccessTime = DateTime.UnixEpoch.AddSeconds(css.AccessTime).ToLocalTime();
```

rjmholt replied:
And below

rjmholt commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -644,21 +988,39 @@ internal static class NativeMethods
-                    public int tm_yday;   /* Day in the year (0-365, 1 Jan = 0) */
-                    public int tm_isdst;  /* Daylight saving time */
+                    /// <summary>Seconds (0-60).</summary>
+                    internal int tm_sec;
```
Any reason for changing these to `internal`? I know it's done elsewhere, but usually my reasoning goes like "If this type were public, should this field be visible?"

rjmholt commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -701,6 +1063,94 @@ internal static UnixTm DateTimeToUnixTm(DateTime date)
                                                         out UInt64 device, out UInt64 inode);
+
+                /// <summary>
+                /// This is a struct from getcommonstat.h in the native library. It's a synthetic construct of the Unix stat
```
Could you reword `synthetic construct`? It initially confused me until I realised we're saying `this struct doesn't occur anywhere like this, but this type represents the maximal union of all structs in the wild`.

JamesWTruher replied:
done!

rjmholt commented on src/System.Management.Automation/engine/TypeTable_Types_Ps1Xml.cs:
```diff
@@ -9198,6 +9198,53 @@ private void Process_Types_Ps1Xml(string filePath, ConcurrentBag<string> errors)
+            #region UnixStat
+
+
+            if (ExperimentalFeature.IsEnabled("PSUnixFileStat"))
```
We could cache this, especially since it's an engine experimental feature, so it can't be toggled once PS has started

adityapatwardhan replied:
Caching not required for perf reasons. I believe changes to all engine experimental features need a restart of PS.

rjmholt commented on src/System.Management.Automation/namespaces/ProviderBase.cs:
```diff
@@ -1862,6 +1862,27 @@ public void WriteInformation(object messageData, string[] tags)
+                if (ExperimentalFeature.IsEnabled("PSUnixFileStat"))
+                {
+                    // Add a commonstat structure to file system objects
+                    if (ProviderInfo.ImplementingType == typeof(Microsoft.PowerShell.Commands.FileSystemProvider))
```
Could squash this into the previous condition with an `&&`

JamesWTruher replied:
done

JamesWTruher commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,285 @@ internal static int NonWindowsGetProcessParentPid(int pid)
+        /// <summary>Unix specific implementations of required functionality.</summary>
         internal static class Unix
         {
+            // Please note that `Win32Exception(Marshal.GetLastWin32Error())`
```
yah - I moved this comment all around to get it through code factor. I can try a different location :)

vexx32 replied:
Maybe we can add it in a `<remarks>` field in an XML comment for the class itself?

JamesWTruher commented on src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:
```diff
@@ -554,27 +554,285 @@ internal static int NonWindowsGetProcessParentPid(int pid)
+                    foreach (StatMask permission in permissions)
+                    {
+                        // determine whether we are setuid, sticky, or the usual rwx.
+                        if ((Mode & (int)permission) == (int)permission)
```
all the examples I could find (and I only looked for a couple) they do something similar to what I have using if/else (freebsd, gnu)

JamesWTruher committed:
Feedback changes from rjmholt

Move comment around to a hopefully acceptable location
Change language which describes the new common stat structure.

adityapatwardhan commented on src/System.Management.Automation/namespaces/ProviderBase.cs:
```diff
@@ -1862,6 +1862,24 @@ public void WriteInformation(object messageData, string[] tags)
+                    }
+                    catch
+                    {
+                        result.AddOrSetProperty("UnixStat", null);
```
use named parameter in place of `null`

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1:
```diff
@@ -0,0 +1,123 @@
+    # if PSUnixFileStat is converted from an experimental feature, these tests will need to be changed
+    BeforeAll {
+        $experimentalFeatureName = "PSUnixFileStat"
+        $skipTest = -not $EnabledExperimentalFeatures.Contains($experimentalFeatureName)
```
Consider setting PSDefaultParameterValue for `Skip`

JamesWTruher replied:
done

adityapatwardhan commented on src/System.Management.Automation/namespaces/ProviderBase.cs:
```diff
@@ -1862,6 +1862,24 @@ public void WriteInformation(object messageData, string[] tags)
+                        var commonStat = Platform.Unix.GetLStat(path);
+                        result.AddOrSetProperty("UnixStat", commonStat);
+                    }
+                    catch
```
Do you want to catch a specific exception here?

JamesWTruher replied:
no, any problem means that we can't add the property, so don't. I'll add a comment

JamesWTruher replied:
added a comment

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/UnixStat.Tests.ps1:
```diff
@@ -0,0 +1,123 @@
+            $group = (/bin/ls -ld $testFile).split(" ",[System.StringSplitOptions]"RemoveEmptyEntries")[3]
+            $i.Group | Should -Be $Group
+        }
+
```
extra line

JamesWTruher replied:
fixed

JamesWTruher committed:
Add a comment in ProviderBase to explain why all exceptions are caught when adding UnixStat.

Fix a couple of small things in the tests, change to use PSDefaultParameterValue rather than using individual skip parameters

JamesWTruher committed:
Merge branch 'master' into CommonStat01
